fs: compaction stats

wmitros opened the pull request:
Commits implementing calculations of the size of metadata log, compacted metadata log, and amount+layout of up-to-date data in every data cluster.

varqox commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -135,7 +135,136 @@ class metadata_log_bootstrap {
 
     future<> bootstrap_delete_dir_entry();
 
+<<<<<<< HEAD
```
Please fix conflict

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,70 @@
+namespace seastar::fs {
+
+// Each cluster_data_vec corresponds to an inode_data_vec
+struct cluster_data_vec {
```
Consider moving cluster_data_vec into cluster_info struct.

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,70 @@
+    file_range data_range;
+};
+
+struct cluster_info {
```
Maybe we could change cluster_info to be a class and make `valid_data_size`, `read_only`, and `data` private? We don't want to modify those fields manually because that could leave the struct in an invalid state. Also, we would like to have some read-only getters for those private fields.

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,70 @@
+    void remove_data(disk_offset_t disk_offset, file_range remove_range) {
+        auto data_vec = std::move(data.extract(--data.upper_bound(disk_offset)).mapped());
+        auto inode = data_vec.data_owner;
+        auto left = data_vec.data_range.beg;
+        auto right = data_vec.data_range.end;
```
```suggestion
        auto [left, right] = data_vec.data_range;
```
You can use structured bindings.

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,70 @@
+        auto right = data_vec.data_range.end;
+
+        if (left < remove_range.beg) {
+            data.emplace(disk_offset + (left - remove_range.beg), cluster_data_vec {inode, {left, remove_range.beg}});
```
It works but maybe consider writing it so we don't have negative values while using uints (`left - remove_range.beg`)?

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,70 @@
+            data.emplace(disk_offset + (left - remove_range.beg), cluster_data_vec {inode, {left, remove_range.beg}});
+        }
+        if (remove_range.end < right) {
+            data.emplace(disk_offset + (remove_range.end - remove_range.beg), cluster_data_vec {inode, {remove_range.end, right}});
```
```suggestion
            data.emplace(disk_offset + remove_range.size(), cluster_data_vec {inode, {remove_range.end, right}});
```

tropuq commented on src/fs/metadata_log.hh:
```diff
@@ -68,6 +69,7 @@ class metadata_log {
     // Estimations of metadata log size used in compaction
     size_t _ondisk_log_size = 0;
     size_t _rewrite_log_size = 0;
+    std::map<cluster_id_t, cluster_info> _clusters;
```
We could use unordered_map.

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,76 @@
+    };
+    size_t valid_data_size = 0;
+    bool read_only = false;
+    std::map<disk_offset_t, cluster_data_vec> data;
```
In compaction we will need to read `data` so we will need getter for that. Also `cluster_data_vec` should be public then.

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,76 @@
+namespace seastar::fs {
+
+class cluster_info {
+
```
```suggestion
```

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,76 @@
+    std::map<disk_offset_t, cluster_data_vec> data;
+
+public:
+
```
```suggestion
```

wmitros replied:
now this line separates functions

tropuq commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,76 @@
+        inode_t data_owner;
+        file_range data_range;
+    };
+    size_t valid_data_size = 0;
```
```suggestion
    size_t _valid_data_size = 0;
```
Also apply to other places.

tropuq commented on src/fs/inode_info.hh:
```diff
@@ -84,12 +84,20 @@ struct inode_info {
         template<class Func>
         void cut_out_data_range(file_range range, Func&& cut_data_vec_processor) {
-            static_assert(std::is_invocable_v<Func, inode_data_vec>);
+            static_assert(std::is_invocable_v<Func, inode_data_vec, cut_result>); 
```
```suggestion
            static_assert(std::is_invocable_v<Func, inode_data_vec, cut_result>);
```

tropuq commented on src/fs/metadata_log.cc:
```diff
@@ -134,40 +168,54 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
         inode_data_vec::in_mem_data {std::move(data)}
     };
-
+    size_t log_increase = ondisk_entry_size(ondisk_small_write_header{0,0,static_cast<uint16_t>(data.size()),0});
```
`ondisk_entry_size(ondisk_small_write_header{0,0,static_cast<uint16_t>(data.size()),0});` doesn't look good. Try to simplify those kinds of calls. Maybe some set of functions like `ondisk_entry_size<ondisk_small_write_header>(data.size());` would be better?

psarna replied:
since we call this several times, it indeed makes sense to have a helper function

tropuq commented on src/fs/metadata_log_operations/write.hh:
```diff
@@ -180,6 +180,7 @@ private:
                     if (buff_bytes_left <= SMALL_WRITE_THRESHOLD) {
                         // TODO: add wasted buff_bytes_left bytes for compaction
                         // No space left in the current to_disk_buffer for medium write - allocate a new buffer
+                        _metadata_log.next_data_cluster();
```
You should call `finished_writing()` for clusters with big writes.

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +100,51 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+        int new_writes_cnt = 0;
+        switch (cr) {
+        case CR::BOTH:
+            new_writes_cnt++;
```
if the implementation falls through, please add an explicit `// fall-through` comment to avoid confusion

varqox replied:
C++17 seems to have attribute for that `[[fallthrough]]` (https://en.cppreference.com/w/cpp/language/attributes/fallthrough)

wmitros replied:
I will change it to ifs; it will be easier to read and shorter

varqox replied:
IMO, switch over enum is better than ifs, because compiler will warn if this enum changes. With ifs you don't have such advantage.

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +100,51 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size(ondisk_small_write_header{});
+            },
+            [&](inode_data_vec::on_disk_data& disk_data) {
+                // TODO: Diffrenciate between large write and large write without mtime
```
*DifferenTiate

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +100,51 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+    _ondisk_log_size += log_increase;
+    _rewrite_log_size += log_increase;
+    // During rewrite we'll start with a truncate setting the file size to remember trailing holes
+    if (not is_directory)
```
following Avi's requests, let's use good old `!` instead of `not` in new code, so there'll be less to fix later

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -210,9 +257,34 @@ void metadata_log::memory_only_add_dir_entry(inode_info::directory& dir, inode_t
+    size_t log_adjust = ondisk_entry_size(ondisk_create_inode_as_dir_entry_header{{}, 0, static_cast<uint16_t>(entry_name.size())}) -
+        ondisk_entry_size(ondisk_create_inode{}) -
+        ondisk_entry_size(ondisk_add_dir_entry_header{0, 0, static_cast<uint16_t>(entry_name.size())});
+    // We'll need to adjust ondisk log size because we'll chang it in following calls
```
changE

psarna reviewed (commented):
Semantically good, I only left some nitpicks.

tropuq commented on src/fs/metadata_disk_entries.hh:
```diff
@@ -204,5 +204,43 @@ constexpr size_t ondisk_entry_size(const ondisk_delete_dir_entry_header& entry)
+constexpr size_t ondisk_entry_size(size_t data_size) noexcept;
+template<>
+constexpr size_t ondisk_entry_size<ondisk_small_write_header>(size_t data_size) noexcept {
+    return sizeof(ondisk_type) + sizeof(ondisk_small_write_header
```
```suggestion
    return sizeof(ondisk_type) + ondisk_entry_size(ondisk_small_write_header
```
Apply in other places.

tropuq commented on src/fs/metadata_disk_entries.hh:
```diff
@@ -204,5 +204,43 @@ constexpr size_t ondisk_entry_size(const ondisk_delete_dir_entry_header& entry)
+template<>
+constexpr size_t ondisk_entry_size<ondisk_create_inode_as_dir_entry_header>(size_t data_size) noexcept {
+    return sizeof(ondisk_type) + sizeof(ondisk_create_inode_as_dir_entry_header
+        {0, 0, static_cast<decltype(ondisk_create_inode_as_dir_entry_header::entry_name_length)>(data_size), 0});
```
```suggestion
        {{0, 0, {0, 0, 0, 0, 0, 0}}, 0, static_cast<decltype(ondisk_create_inode_as_dir_entry_header::entry_name_length)>(data_size)});
```

varqox commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,81 @@
+
+#include "fs/cluster.hh"
+#include "fs/inode.hh"
+#include "fs/units.hh"
```
```suggestion
#include "fs/units.hh"

```

varqox commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,81 @@
+    std::map<disk_offset_t, cluster_data_vec> _data;
+
+public:
+    constexpr const decltype(_data)& get_data() const & noexcept {
```
```suggestion
    constexpr const decltype(_data)& get_data() const noexcept {
```
This is not needed if it is the only overload of this method.

varqox replied:
For the sake of readability when navigating the code you could write `std::map<disk_offset_t, cluster_data_vec>` instead of decltype because this type is not such a pain to write and read. Furthermore, since you return const reference, if the type changes compiler will give error or at least warning about returning reference to temporary if implicit conversion happens.

varqox commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,81 @@
+
+namespace seastar::fs {
+
+class cluster_info {
```
```suggestion
class cluster_content_info {
```
Current name is not expressive enough imo, but I cannot come up with a very good name. Something like "cluster_zapełnienie_info" would be great but English has no good word meaning "zapełnienie" or something we want. Maybe "cluster_fragmentation_info" or also replacing "info" with "statistics" would help?

varqox replied:
Also name should say that it is only used by data logs, e.g. adding a prefix `data_` to the name would be sufficient for me.

varqox commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,81 @@
+        _valid_data_size += data_range.size();
+    }
+
+    void remove_data(disk_offset_t disk_offset, file_range remove_range) {
```
I would like to see, at least a comment what this function does, because implementation is not obvious about what exactly happens.

varqox replied:
If you don't want it doesn't have to be as expressive as this one:
https://github.com/psarna/seastar/blob/47620f096e8b41d295e97cf644727d60496b248f/src/fs/inode_info.hh#L108-L112
But it should clearly state what happens here.

varqox commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,81 @@
+    void remove_data(disk_offset_t disk_offset, file_range remove_range) {
+        auto data_vec = std::move(_data.extract(--_data.upper_bound(disk_offset)).mapped());
+        auto inode = data_vec.data_owner;
+        auto [left, right] = data_vec.data_range;
```
maybe `auto dr = data_vec.data_range` and use dr.beg and dr.end below?

varqox commented on src/fs/cluster_info.hh:
```diff
@@ -0,0 +1,81 @@
+        }
+        if (remove_range.end < right) {
+            _data.emplace(disk_offset + remove_range.size(), cluster_data_vec {inode, {remove_range.end, right}});
+        }
```
Maybe using `auto left = intersection(data_vec.data_range, {0, remove_range.beg});` and later checking for `!left.is_empty()` would be more clear, less error prone and less low-level range manipulation?

varqox commented on src/fs/inode_info.hh:
```diff
@@ -138,16 +146,21 @@ struct inode_info {
                 if (not right.data_range.is_empty()) {
                     data.emplace(right.data_range.beg, std::move(right));
+                } else {
+                    cr = cut_result::RIGHT;
```
```suggestion
                    cr = cut_result::LEFT;
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -138,16 +146,21 @@ struct inode_info {
                 if (not left.data_range.is_empty()) {
                     data.emplace(left.data_range.beg, std::move(left));
+                } else {
+                    cr = cut_result::LEFT;
```
```suggestion
                    cr = cut_result::RIGHT;
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -84,12 +84,20 @@ struct inode_info {
             return (data.empty() ? 0 : data.rbegin()->second.data_range.end);
         }
 
+        // Represents which data vectors are reconstructed after removing intersection
```
```suggestion
        // Tells which parts of data vector (relative to intersection) are reconstructed after removing intersection
```
I think it is better this way.

tropuq commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+    _ondisk_log_size += log_increase;
+    _rewrite_log_size += log_increase;
+    // During rewrite we'll start with a truncate setting the file size to remember trailing holes
+    if (!is_directory)
+        _rewrite_log_size += ondisk_entry_size<ondisk_truncate>();
```
```suggestion
    if (!is_directory) {
        _rewrite_log_size += ondisk_entry_size<ondisk_truncate>();
    }
```

varqox commented on src/fs/metadata_disk_entries.hh:
```diff
@@ -204,5 +204,43 @@ constexpr size_t ondisk_entry_size(const ondisk_delete_dir_entry_header& entry)
+template<>
+constexpr size_t ondisk_entry_size<ondisk_small_write_header>(size_t data_length) noexcept {
+    using header = ondisk_small_write_header;
+    return ondisk_entry_size(header{.length = static_cast<decltype(header::length)>(data_length)});
```
Looks good, but I would like to see ondisk_entry_size(ondisk_...) use ondisk_entry_size<ondisk_...>(...) rather than the other way.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            new_writes_cnt = 1;
+        } else {
+            new_writes_cnt = 0;
+        }
```
```suggestion
        int new_writes_cnt = [&] {
            switch (cr) {
            case CR::BOTH: return 2;
            case CR::LEFT: return 1;
            case CR::RIGHT: return 1;
            case CR::NONE: return 0;
            }
        }();
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            [&](inode_data_vec::in_mem_data& mem) {
+                size_t log_decrease = ondisk_entry_size<ondisk_small_write_header>(data_vec.data_range.size());
+                _rewrite_log_size -= log_decrease;
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size<ondisk_small_write_header>(0);
```
This is tricky, especially when reader notices "ondisk_entry_size<ondisk_small_write_header>(0)".

varqox replied:
It is a symptom that it could be done in a more elegant way, eg. cut_out_data_range could give you inode_data_vec left_remainder, inode_data_vec removed, inode_data_vec right_remainder, where (left|right)_remainder would be empty if they are removed. This way you have an exact inode_data_vec and don't have to do tricky calculations every time, just calling `ondisk_entry_size<ondisk_small_write_header>(left_remainder) + ...`. The only problem is that you have to calculate the size of the original (before cutting) data_vec, but this can be simply done in the same way: providing it as parameter to the callback, so that there are 4: original, left_remainder, removed, right_remainder.
What do you think?

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                    ondisk_entry_size<ondisk_large_write>() :
+                    ondisk_entry_size<ondisk_medium_write>();
+                _rewrite_log_size -= log_decrease;
+                // Remaining data is smaller than original so it's less than large write
```
```suggestion
                // Remaining data is smaller than original so it is not a large write anymore
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                _rewrite_log_size -= log_decrease;
+                // Remaining data is smaller than original so it's less than large write
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size<ondisk_medium_write>();
+                _clusters[offset_to_cluster_id(disk_data.device_offset, _cluster_size)].remove_data(disk_data.device_offset, data_vec.data_range);
```
`_clusters` this is not a good name it should be corresponding to the name of the struct cluster_info I have elaborated earlier.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            },
+            [&](inode_data_vec::on_disk_data& disk_data) {
+                // TODO: Differentiate between large write and large write without mtime
+                size_t log_decrease = data_vec.data_range.size() == _cluster_size ?
+                    ondisk_entry_size<ondisk_large_write>() :
+                    ondisk_entry_size<ondisk_medium_write>();
+                _rewrite_log_size -= log_decrease;
+                // Remaining data is smaller than original so it's less than large write
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size<ondisk_medium_write>();
```
The same as holds here. (see comment to `inode_data_vec::in_mem_data&` case)

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            new_writes_cnt = 0;
+        }
+        std::visit(overloaded {
+            [&](inode_data_vec::in_mem_data& mem) {
```
```suggestion
            [&](const inode_data_vec::in_mem_data&) {
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size<ondisk_medium_write>();
+                _clusters[offset_to_cluster_id(disk_data.device_offset, _cluster_size)].remove_data(disk_data.device_offset, data_vec.data_range);
+            },
+            [&](inode_data_vec::hole_data&) {
```
```suggestion
            [&](const inode_data_vec::hole_data&) {
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                _rewrite_log_size -= log_decrease;
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size<ondisk_small_write_header>(0);
+            },
+            [&](inode_data_vec::on_disk_data& disk_data) {
```
```suggestion
            [&](const inode_data_vec::on_disk_data& disk_data) {
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
 inode_info& metadata_log::memory_only_create_inode(inode_t inode, bool is_directory, unix_metadata metadata) {
     assert(_inodes.count(inode) == 0);
+    size_t log_increase = ondisk_entry_size<ondisk_create_inode>();
+    _ondisk_log_size += log_increase;
```
_ondisk_log_size is redundant (too high accuracy, maybe even counting not what we want) imo. What is enough is the number of clusters that metadata log takes.

varqox replied:
If you want to stay with _ondisk_log_size, does it accounts for added padding when flushing metadata log?

wmitros replied:
We can't calculate precisely the number of clusters in the compacted metadata log nor account for added padding in it, so for the sake of comparison we're not doing it for the written metadata log too.

varqox replied:
Yeah, that is true (but I think it is achievable although hard to take padding into account, regarding _ondisk_log_size).
Anyway, calculating the number of clusters that metadata log takes is trivial, this takes padding into account and is accurate (because this is the space that metadata log occupies). Padding is problematic to foresee -- I totally agree. But that is the margin of error that should be negligible.
_ondisk_log_size is not a good comparison because scenarios like this exist: suppose that I fill the log with 10000 entries of creating inode (that is really a small entry) and after adding each, I trigger the flush_log(), so that the padding is added. This way metadata log will take about 100 times more space than _ondisk_log_size is indicating (this number is for alignment == 4096). This is really bad that, because we could reclaim a lot of space using compaction, but this metric would indicate that it is not worth compacting because _ondisk_log_size == _rewrite_log_size (more or less).

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -134,40 +166,54 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
 
     std::visit(overloaded {
-        [](const inode_info::directory& dir) {
+        [](inode_info::directory& dir) {
```
```suggestion
        [](const inode_info::directory& dir) {
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -134,40 +166,54 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
+                0,
+                std::numeric_limits<decltype(file_range::end)>::max()
+            });
+            ondisk_entry_size<ondisk_truncate>();
```
```suggestion
            _rewrite_log_size -= ondisk_entry_size<ondisk_truncate>();
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -134,40 +166,54 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
 void metadata_log::memory_only_disk_write(inode_t inode, file_offset_t file_offset, disk_offset_t disk_offset,
         size_t write_len) {
+    assert(write_len <= _cluster_size);
+    size_t log_increase = write_len == _cluster_size ?
+        ondisk_entry_size<ondisk_large_write>() :
+        ondisk_entry_size<ondisk_medium_write>();
```
```suggestion
    size_t log_increase = (write_len == _cluster_size ?
        ondisk_entry_size<ondisk_large_write>() :
        ondisk_entry_size<ondisk_medium_write>());
```
I think that `=` and `==` that are near one another may look deceptive. And it is error prone, consider:
```
size_t log_increase = write_len = _cluster_size ?
```
I would work (without warnings?).

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -210,9 +258,34 @@ void metadata_log::memory_only_add_dir_entry(inode_info::directory& dir, inode_t
+
+inode_info& metadata_log::memory_only_create_inode_as_dir_entry(inode_t entry_inode, bool is_directory, unix_metadata metadata,
+        inode_info::directory& dir, std::string entry_name) {
+    size_t log_adjust = ondisk_entry_size<ondisk_create_inode_as_dir_entry_header>(entry_name.size()) -
+        ondisk_entry_size<ondisk_create_inode>() -
+        ondisk_entry_size<ondisk_add_dir_entry_header>(entry_name.size());
```
Are we ok with the unsigned overflows here? (they are not UB)

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -222,6 +295,24 @@ void metadata_log::memory_only_delete_dir_entry(inode_info::directory& dir, std:
+    }
+}
+
+void metadata_log::memory_only_delete_inode_after_dir_entry(inode_t entry_inode) {
+    size_t log_adjust = ondisk_entry_size<ondisk_delete_inode_and_dir_entry_header>(0) -
+        ondisk_entry_size<ondisk_delete_inode>() - ondisk_entry_size<ondisk_delete_dir_entry_header>(0);
+    memory_only_delete_inode(entry_inode);
+    _ondisk_log_size += log_adjust;
+}
```
What is that?
after `after` and before `_dir_entry` should be a verb (adding?, deleting?).

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -222,6 +295,24 @@ void metadata_log::memory_only_delete_dir_entry(inode_info::directory& dir, std:
+}
+
+void metadata_log::memory_only_delete_inode_after_dir_entry(inode_t entry_inode) {
+    size_t log_adjust = ondisk_entry_size<ondisk_delete_inode_and_dir_entry_header>(0) -
```
I don't like these zeros, it should not be like that.

varqox commented on src/fs/metadata_log_operations/unlink_or_remove_file.hh:
```diff
@@ -150,7 +150,7 @@ class unlink_or_remove_file_operation {
                 return make_exception_future(no_more_space_exception());
             case metadata_log::append_result::APPENDED:
                 _metadata_log.memory_only_delete_dir_entry(*_dir_info, _entry_name);
-                _metadata_log.memory_only_delete_inode(_entry_inode);
+                _metadata_log.memory_only_delete_inode_after_dir_entry(_entry_inode);
```
I think it should be done as call to one function if you need more context there. Just like memory_only_create_inode_as_dir_entry() is done.

wmitros replied:
it is done this way because these calls are separated in the corresponding function in bootstrap

varqox replied:
I think it can be rewritten so that the operations there are called one after another, so that it will be possible to merge them.

varqox replied:
I think that commit that rewrites it (this bootstraping method) should go as 'FIXUP:' in this series, so that i can land in zpp_fs before this pull request. And another on top of that merges these functions together and this commit will stay here in this pull request. So 1. FIXUP commit that rewrites bootstraping method so that these calls are alongside each other. 2. commit that merges these calls into one.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -222,6 +295,24 @@ void metadata_log::memory_only_delete_dir_entry(inode_info::directory& dir, std:
+    _ondisk_log_size += log_adjust;
+}
+
+void metadata_log::finish_data_cluster(cluster_id_t cluster_id) {
```
```suggestion
void metadata_log::finished_writing_cluster(cluster_id_t cluster_id) {
```

varqox commented on src/fs/metadata_log.hh:
```diff
@@ -65,6 +66,11 @@ class metadata_log {
+    // Estimations of metadata log size used in compaction
+    size_t _ondisk_log_size = 0;
+    size_t _rewrite_log_size = 0;
+    std::unordered_map<cluster_id_t, cluster_info> _clusters;
```
Maybe instead of keeping record if cluster is writable or not in the cluster itself, splitting this into two:
- _writable_clusters
- _read_only_clusters
?
Ofc, that is just an idea not a "requested change".

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,82 @@
+    std::map<disk_offset_t, cluster_data_vec> _data;
+
+public:
+    constexpr const std::map<disk_offset_t, cluster_data_vec>& get_data() const noexcept {
```
```suggestion
    const std::map<disk_offset_t, cluster_data_vec>& get_data() const noexcept {
```
There is no use in making it constexpr, as std::map and thus the whole data_cluster_contents_info cannot be constructed in compile time. Also if adding constexpr, it should be added in every place applicable -- the other methods that may be evaluated in compile time. But again, here it serves no purpose.

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,82 @@
+        return _valid_data_size == 0;
+    }
+
+    void finished_writing() noexcept {
```
```suggestion
    void make_read_only() noexcept {
```
Maybe this would be better? finished_writing is linked to the specific use case we have and we can easily give it a more general name.

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,82 @@
+    };
+
+private:
+    size_t _valid_data_size = 0;
```
I would change `valid` to either `vital`, `needful` or `active`. Valid indicates correctness, but we want to indicate being up-to-date and still relevant.

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,82 @@
+        _data.emplace(disk_offset, cluster_data_vec {inode, data_range});
+        _valid_data_size += data_range.size();
+    }
+    // Removes data_vec stored at disk_offset and adds remaining data_
```
Comment seems to be out-of-date

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,82 @@
+    void add_data(disk_offset_t disk_offset, inode_t inode, file_range data_range) noexcept {
+        _data.emplace(disk_offset, cluster_data_vec {inode, data_range});
+        _valid_data_size += data_range.size();
+    }
```
```suggestion
    }

```

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,82 @@
+    void cut_data(disk_offset_t disk_offset, file_range former_range, file_range new_left_range, file_range new_right_range) {
+        auto data_vec_nh = _data.extract(disk_offset);
+        auto inode = data_vec_nh.mapped().data_owner;
+        if (new_left_range.size() > 0) {
```
```suggestion
        if (!new_left_range.is_empty()) {
```

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,82 @@
+            data_vec_nh.mapped() = {inode, new_left_range};
+            _data.insert(std::move(data_vec_nh));
+        }
+        if (new_right_range.size() > 0) {
```
```suggestion
        if (!new_right_range.is_empty()) {
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -99,55 +99,46 @@ struct inode_info {
 
-                // Overlaps => cut it, possibly into two parts:
+                // Overlaps => find leading(left) and tracing(right) segments of the data_vec
+                // that remain after cutting the intersection,
```
```suggestion
                // that remain after cutting out the intersection,
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -99,55 +99,46 @@ struct inode_info {
+                // finally insert each of left and right unless it's empty
                 // |       data_vec      |
                 //         | cap |
                 // | left  | mid | right |
```
```suggestion
                // | left  | cap | right |
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -99,55 +99,46 @@ struct inode_info {
+                // Overlaps => find leading(left) and tracing(right) segments of the data_vec
+                // that remain after cutting the intersection,
+                // then delete the data_vec that intersects,
+                // finally insert each of left and right unless it's empty
```
I am not sure if this comment is too verbose -- it states what happens below, but in a way that is totally the same as the code below. I think that comments should express _what_ and _why_ rather than _how_ unless the _how_ is not obvious or tricky. Here the _how_ part is apparent if you look at the code below.

varqox commented on src/fs/inode_info.hh:
```diff
@@ -99,55 +99,46 @@ struct inode_info {
                     },
                 }, data_vec.data_location);
 
+                // Process deleted vector
```
```suggestion
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -89,7 +89,7 @@ struct inode_info {
         // data vectors) that will be deleted
         template<class Func>
         void cut_out_data_range(file_range range, Func&& cut_data_vec_processor) {
```
```suggestion
        void cut_out_data_range(file_range range, Func&& data_vec_cutting_callback) {
```
Remember to update all comments.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -23,6 +23,7 @@
 #include "fs/cluster_allocator.hh"
 #include "fs/inode.hh"
 #include "fs/inode_info.hh"
+#include "fs/data_cluster_contents_info.hh"
```
Please sort includes.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                },
+                [&](const inode_data_vec::hole_data&) {
+                },
+            }, right_remains.data_location);
```
Cmon, it is literally copy paste of the above. Make a lambda taking data_location and run in on `left_remains.data_location` and `right_remains.data_location` in these ifs.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+        std::visit(overloaded {
+            [&](const inode_data_vec::in_mem_data&) {
+                size_t log_decrease = ondisk_entry_size<ondisk_small_write_header>(former.data_range.size());
+                _rewrite_log_size -= log_decrease;
```
```suggestion
                _rewrite_log_size -= ondisk_entry_size<ondisk_small_write_header>(former.data_range.size());
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            },
+            [&](const inode_data_vec::hole_data&) {
+            },
+        }, former.data_location);
```
Maybe move that above the above ifs, so that flow is the same as in other places: removing the former, then adding what remains.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            [&](const inode_data_vec::hole_data&) {
+            },
+        }, former.data_location);
+
```
```suggestion
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                size_t log_decrease = (former.data_range.size() == _cluster_size ?
+                    ondisk_entry_size<ondisk_large_write>() :
+                    ondisk_entry_size<ondisk_medium_write>());
+                _rewrite_log_size -= log_decrease;
```
```suggestion
                _rewrite_log_size -= (former.data_range.size() == _cluster_size ?
                    ondisk_entry_size<ondisk_large_write>() :
                    ondisk_entry_size<ondisk_medium_write>());
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                    ondisk_entry_size<ondisk_large_write>() :
+                    ondisk_entry_size<ondisk_medium_write>());
+                _rewrite_log_size -= log_decrease;
+                // Remaining data is smaller than original so it is not a large write anymore
```
I would add an assert for that, just to make sure that it is used correctly -- caller does not have a bug.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
 inode_info& metadata_log::memory_only_create_inode(inode_t inode, bool is_directory, unix_metadata metadata) {
     assert(_inodes.count(inode) == 0);
+    _rewrite_log_size += ondisk_entry_size<ondisk_create_inode>();
+    // During rewrite we'll start with a truncate setting the file size to remember trailing holes
```
```suggestion
    // During rewrite we'll start with truncating the file to its final size to remember trailing holes
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -136,37 +183,48 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
+                0,
+                std::numeric_limits<decltype(file_range::end)>::max()
+            });
+            _rewrite_log_size -= ondisk_entry_size<ondisk_truncate>();
```
```suggestion
            _rewrite_log_size -= ondisk_entry_size<ondisk_truncate>(); // See memory_only_create_inode() for why it is here
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -136,37 +183,48 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
+        ondisk_entry_size<ondisk_large_write>() :
+        ondisk_entry_size<ondisk_medium_write>());
+    cluster_id_t cluster_id = offset_to_cluster_id(disk_offset, _cluster_size);
+    _writable_data_clusters[cluster_id].add_data(disk_offset, inode, {file_offset, file_offset + write_len});
```
```suggestion
    _writable_data_clusters.at(cluster_id).add_data(disk_offset, inode, {file_offset, file_offset + write_len});
```
Use at() (preferably i would like to have version of at() that instead of throwing an exception calls std::terminate -- asserts that the key exists). operator[] creates element in the map, so if the write is misused here, there will be no error.

varqox replied:
Please look at the comment about at() that is somewhere above.

psarna commented on src/fs/metadata_log.hh:
```diff
@@ -65,6 +66,13 @@ class metadata_log {
+    size_t _rewrite_log_size = 0;
+    std::unordered_map<cluster_id_t, data_cluster_contents_info*> _data_cluster_contents_info_map;
+    std::map<cluster_id_t, data_cluster_contents_info> _writable_data_clusters;
+    std::map<cluster_id_t, data_cluster_contents_info> _read_only_data_clusters;
```
Do we need ordering here? I didn't find any indication that we do, so `unordered_map` may be more efficient here as well.

psarna commented on src/fs/metadata_log_operations/write.hh:
```diff
@@ -185,6 +185,7 @@ private:
                             // TODO: maybe we should return partial write instead of exception?
                             return make_exception_future<bool_class<stop_iteration_tag>>(no_more_space_exception());
                         }
+                        _metadata_log.finish_writing_cluster(offset_to_cluster_id(_metadata_log._curr_data_writer->current_disk_offset(), _metadata_log._cluster_size));
```
nitpick: let's split this turboline into 2 shorter lines

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,65 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+                    ondisk_entry_size<ondisk_medium_write>());
+                _rewrite_log_size -= log_decrease;
+                // Remaining data is smaller than original so it is not a large write anymore
+                _data_cluster_contents_info_map[offset_to_cluster_id(disk_data.device_offset, _cluster_size)]->cut_data(
```
```suggestion
                _data_cluster_contents_info_map.at(offset_to_cluster_id(disk_data.device_offset, _cluster_size))->cut_data(
```
Use at(), because operator[] creates element if it does not exist and that is not what we want here -- for us it is a bug. Although an assert is preferable for this, at() throws exception and there is no version of at() that would assert that the key exists.

varqox replied:
```suggestion
                _data_cluster_contents_info_map.find(offset_to_cluster_id(disk_data.device_offset, _cluster_size))->second->cut_data(
```
That is the second option that I would like better, but there is no checking here so it is better to write it in that way:
```suggestion
                auto it = _data_cluster_contents_info_map.find(offset_to_cluster_id(disk_data.device_offset, _cluster_size));
                assert(it != _data_cluster_contents_info_map.end());
                it->second->cut_data(
```
But the most preferable would be some version of at that asserts instead of throws, eg.
```suggestion
                our_at(_data_cluster_contents_info_map, offset_to_cluster_id(disk_data.device_offset, _cluster_size))->cut_data(
```

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -221,6 +287,19 @@ void metadata_log::memory_only_delete_dir_entry(inode_info::directory& dir, std:
     --entry_it->second.directories_containing_file;
     dir.entries.erase(it);
+
+    if (entry_it->second.directories_containing_file == 0) {
```
nitpick: If this is a rather unexpected state, we can consider wrapping it with `__buildin_expect(false)`

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -221,6 +287,19 @@ void metadata_log::memory_only_delete_dir_entry(inode_info::directory& dir, std:
+}
+
+void metadata_log::finish_writing_cluster(cluster_id_t cluster_id) {
+    _writable_data_clusters[cluster_id].finished_writing();
```
```suggestion
    _writable_data_clusters.at(cluster_id).finished_writing();
```
Same here.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -255,6 +334,7 @@ metadata_log::flush_result metadata_log::schedule_flush_of_curr_cluster_and_chan
 
     auto append_res = _curr_cluster_buff->append(ondisk_next_metadata_cluster {*next_cluster});
     assert(append_res == metadata_to_disk_buffer::APPENDED);
+    _log_cluster_count++;
```
```suggestion
    ++_log_cluster_count;
```
IMO this looks better, but I will respect if you prefer otherwise.

varqox replied:
Please either apply this or resolve this.

varqox commented on src/fs/metadata_log.hh:
```diff
@@ -65,6 +66,13 @@ class metadata_log {
 
+    // Estimations of metadata log size used in compaction
+    cluster_id_t _log_cluster_count = 0;
+    size_t _rewrite_log_size = 0;
```
```suggestion
    size_t _rewritten_log_size = 0;
```
or
```suggestion
    size_t _compacted_log_size = 0;
```
What do you think?

varqox reviewed (changes requested):
Regarding the whole drama with at() function, I didn't see explicitly creating entries in _writable_data_clusters and derivatives. It has to be done when allocating cluster for writing and somewhere in bootstrap? Maybe it is too much to change and it is better to create entries on demand, but if it stays that way, we need asserting that the equivalent in _read_only_data_clusters does not exist. Regarding that I did not see inserting to _data_cluster_contents_info_map anywhere and that is certainly a bug, because default constructing a pointer gives nullptr, so creating entry like that:
```
_data_cluster_contents_info_map[...]->do_sth();
```
Is UB and probably will cause SEGV.

varqox commented:
Indeed, the tests show that the code is causing null pointer dereference.

psarna commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    // Removes data_vec stored at disk_offset and adds remaining data_
+    void cut_data(disk_offset_t disk_offset, file_range former_range, file_range new_left_range, file_range new_right_range) {
```
Looks good, but it took me a while to go over some example corner cases and convince myself that there are no obvious bugs. Do we have a unit test for the case when *both* left and right remains, i.e. when we punch a hole in a file? I saw truncate tests, but unless I'm mistaken, they only cover cutting data from the end.

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,95 @@
+
+    bool can_compact() const noexcept {
+        return _read_only;
+    }
```
Since we have distinction between RO and WR clusters managed at the higher abstraction level this functions are redundant (note that `_read_only` field also should be removed).

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    void add_data(disk_offset_t disk_offset, inode_t inode, file_range data_range) noexcept {
+        {
+            auto it =_data.upper_bound(disk_offset);
+            assert(it == _data.end() || it->first >= disk_offset + data_range.size());
+            if (it != _data.begin()) {
+                it--;
+                assert(disk_offset >= it->first + it->second.data_range.size());
+            }
+        }
```
What is this for?

wmitros replied:
this checks if the disk range that will be inserted does not overlap with existing data_vecs

varqox replied:
emplace() returns an iterator. I think that using it would be more efficient and more clear here. Could you try that?

varqox commented on src/fs/inode_info.hh:
```diff
@@ -99,55 +99,42 @@ struct inode_info {
 
-                // Overlaps => cut it, possibly into two parts:
+                // If our range overlaps with current data_vec, we erase it,
+                // and up to 2 parts that didn't intersect are inserted back
```
```suggestion
                // If our range overlaps with current data_vec, we cut out the intersection leaving at most two parts
```
I think this way it sounds better.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,55 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
-    file.cut_out_data_range(range, [](inode_data_vec data_vec) {
-        (void)data_vec; // TODO: for compaction: update used inode_data_vec
+    file.cut_out_data_range(range, [&](const inode_data_vec& former, const inode_data_vec& left_remains, const inode_data_vec& right_remains) {
+        std::visit(overloaded {
```
Please add a comment that you remove the `former` from the `_compacted_log_size`. It is not obvious unless you start reading the code thoroughly.

varqox replied:
Oh it is not so trivial to comment because of calling `cut_data` which is the correct thing to do here imo.

wmitros replied:
Is this suggestion about _compacted_log_size or about data_cluster_contents, because in the first case I don't see how the "_compacted_log_size -= ..." lines could benefit from any comments

varqox replied:
Let me help you
```suggestion
        // Remove former before adding remains
        std::visit(overloaded {
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,55 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            },
+            [&](const inode_data_vec::hole_data&) {
+            },
+        }, former.data_location);
```
```suggestion
        }, former.data_location);

```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,55 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+            },
+        }, former.data_location);
+        auto visit_remains = [&](const inode_data_vec& data_vec) {
+            if (!data_vec.data_range.is_empty()) {
```
You can remove one indentation simply by returning early on negative condition:
```suggestion
            if (data_vec.data_range.is_empty()) {
                return;
            }
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,55 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
+    _compacted_log_size += ondisk_entry_size<ondisk_create_inode>();
+    // During rewrite we'll start with truncating the file to its final size to remember trailing holes
+    if (!is_directory) {
+        _compacted_log_size += ondisk_entry_size<ondisk_truncate>();
```
I'm not sure which placement for the comment is better, what do you think @psarna?
```suggestion
    if (!is_directory) {
        // During rewrite we'll start with truncating the file to its final size to remember trailing holes
        _compacted_log_size += ondisk_entry_size<ondisk_truncate>();
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -136,37 +173,49 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
+    cluster_id_t cluster_id = offset_to_cluster_id(disk_offset, _cluster_size);
+    assert(_read_only_data_clusters.find(cluster_id) == _read_only_data_clusters.end());
+    _writable_data_clusters[cluster_id].add_data(disk_offset, inode, {file_offset, file_offset + write_len});
+    _data_cluster_contents_info_map[cluster_id] = &(_writable_data_clusters[cluster_id]);
```
Parenthesis is not needed after `&`, also we could avoid double lookup simply by taking a reference.
```suggestion
    auto& cc_info = _writable_data_clusters[cluster_id];
    cc_info.add_data(disk_offset, inode, {file_offset, file_offset + write_len});
    _data_cluster_contents_info_map[cluster_id] = &cc_info;
```

tropuq commented on src/fs/metadata_log.hh:
```diff
@@ -201,6 +209,8 @@ private:
     void memory_only_add_dir_entry(inode_info::directory& dir, inode_t entry_inode, std::string entry_name);
     void memory_only_delete_dir_entry(inode_info::directory& dir, std::string entry_name);
 
+    void finish_writing_cluster(cluster_id_t cluster_id);
```
Maybe `finish_writing_data_cluster` to make it more clear that it is used only with data clusters?

tropuq replied:
It was resolved without any action or comment.

wmitros replied:
my bad, applied now

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -209,6 +257,15 @@ void metadata_log::memory_only_add_dir_entry(inode_info::directory& dir, inode_t
     ++it->second.directories_containing_file;
+
+    if (it->second.directories_containing_file == 1) {
+        // We know that we'll write ondisk_create_inode_as_dir_entry_header(instead of ondisk_create_inode and ondisk_add_dir_entry)
```
```suggestion
        // We know that we'll write ondisk_create_inode_as_dir_entry_header (instead of ondisk_create_inode and ondisk_add_dir_entry)
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -221,6 +278,20 @@ void metadata_log::memory_only_delete_dir_entry(inode_info::directory& dir, std:
+}
+
+void metadata_log::finish_writing_cluster(cluster_id_t cluster_id) {
+    assert(_writable_data_clusters.count(cluster_id) == 1 && _read_only_data_clusters.count(cluster_id) == 0);
+    _writable_data_clusters.at(cluster_id).make_read_only();
+    _read_only_data_clusters.insert(std::move(_writable_data_clusters.extract(cluster_id)));
```
```suggestion
    auto nh = _writable_data_clusters.extract(cluster_id);
    assert(!nh.empty());
    auto insert_res = _read_only_data_clusters.insert(std::move(nh));
    assert(insert_res.inserted);
```

varqox commented on src/fs/metadata_log.hh:
```diff
@@ -26,6 +26,7 @@
 #include "fs/cluster_writer.hh"
 #include "fs/inode.hh"
 #include "fs/inode_info.hh"
+#include "fs/data_cluster_contents_info.hh"
 #include "fs/metadata_disk_entries.hh"
 #include "fs/metadata_to_disk_buffer.hh"
 #include "fs/units.hh"
```
Please sort includes.

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -52,25 +51,18 @@ public:
-            }
-        }
-        _data.emplace(disk_offset, cluster_data_vec {inode, data_range});
+        auto [it, succ] = _data.emplace(disk_offset, cluster_data_vec {inode, data_range});
+        assert(succ);
```
It is more idiomatic to name it this way
```suggestion
        auto [it, inserted] = _data.emplace(disk_offset, cluster_data_vec {inode, data_range});
        assert(inserted);
```

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -52,25 +51,18 @@ public:
+        assert(succ);
         _up_to_date_data_size += data_range.size();
+        if (it != _data.begin()) {
+            it--;
```
@psarna what is your opinion about using pre-increment instead of post-increment for iterator types? This may be relevant in the debug build where the compiler does not optimize it away.

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -52,25 +51,18 @@ public:
         _up_to_date_data_size += data_range.size();
+        if (it != _data.begin()) {
+            it--;
+            assert(disk_offset >= it->first + it->second.data_range.size());
```
Anyway, you decrease iterator and the increase it up, you could just use `std::prev()` to get the value you want without going back and forth.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -105,6 +105,7 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
 
 void metadata_log::cut_out_data_range(inode_info::file& file, file_range range) {
     file.cut_out_data_range(range, [&](const inode_data_vec& former, const inode_data_vec& left_remains, const inode_data_vec& right_remains) {
+        // Former is a data_vec that intersected with range, left_remains and right_remains are parts of it that didn't
```
This comment is redundant for me.

wmitros replied:
https://github.com/psarna/seastar/pull/96#discussion_r417476741

varqox replied:
But try to compare these comments -- they tell different things.

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,83 @@
+        auto [it, inserted] = _data.emplace(disk_offset, cluster_data_vec {inode, data_range});
+        assert(inserted);
+        _up_to_date_data_size += data_range.size();
+        auto pv = std::prev(it, 1);
```
1 is redundant.
```suggestion
        auto pv = std::prev(it);
```

varqox commented on src/fs/data_cluster_contents_info.hh:
```diff
@@ -0,0 +1,83 @@
+        _up_to_date_data_size += data_range.size();
+        auto pv = std::prev(it, 1);
+        assert(it == _data.begin() || disk_offset >= pv->first + pv->second.data_range.size());
+        auto nx = std::next(it, 1);
```
```suggestion
        auto nx = std::next(it);
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -85,11 +85,11 @@ struct inode_info {
         template<class Func>
-        void cut_out_data_range(file_range range, Func&& cut_data_vec_processor) {
-            static_assert(std::is_invocable_v<Func, inode_data_vec>);
+        void cut_out_data_range(file_range range, Func&& data_vec_cutting_callback) {
```
A description of what data_vec_cutting_callback parameters are should be provided in the comment above.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -100,18 +100,58 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
-    file.cut_out_data_range(range, [](inode_data_vec data_vec) {
-        (void)data_vec; // TODO: for compaction: update used inode_data_vec
+    file.cut_out_data_range(range, [&](const inode_data_vec& former, const inode_data_vec& left_remains, const inode_data_vec& right_remains) {
+        std::visit(overloaded {
+        // Remove former before adding remains
```
```suggestion
        // Remove former before adding remains
        std::visit(overloaded {
```

varqox commented:
There are some TODOs about compaction stats, please fix every adequate one:
```
src/fs/metadata_log.cc:328:        // TODO: ^ add protection from it and take it into account during compaction
src/fs/metadata_log.cc:595:// TODO: Maybe move compaction to another module
src/fs/metadata_log.hh:163:    // TODO: for compaction: keep some set(?) of inode_data_vec, so that we can keep track of clusters that have lowest
src/fs/metadata_log.hh:165:    // TODO: for compaction: keep estimated metadata log size (that would take when written to disk) and
src/fs/metadata_log_operations/read.hh:78:                // TODO: for compaction: mark that clusters shouldn't be moved to _cluster_allocator before that read ends
src/fs/metadata_log_operations/write.hh:181:                        // TODO: add wasted buff_bytes_left bytes for compaction
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -136,37 +176,50 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
-            // TODO: for compaction: update used inode_data_vec
+        [&](inode_info::file& file) {
+            cut_out_data_range(file, {
+                0,
+                std::numeric_limits<decltype(file_range::end)>::max()
```
Please use designated initializers.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -136,37 +176,50 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
+        ondisk_entry_size<ondisk_medium_write>());
+    cluster_id_t cluster_id = offset_to_cluster_id(disk_offset, _cluster_size);
+    assert(_read_only_data_clusters.find(cluster_id) == _read_only_data_clusters.end());
+    auto& cc_info = _writable_data_clusters[cluster_id];
+    cc_info.add_data(disk_offset, inode, {file_offset, file_offset + write_len});
+    _data_cluster_contents_info_map[cluster_id] = &cc_info;
```
We could do an optimization here that `_data_cluster_contents_info_map[cluster_id]` is not updated unless `_writable_data_clusters[cluster_id]` created the element. If you change it `try_emplace` will do the job well:
```
auto [cc_info_it, created] = _writable_data_clusters.try_emplace(clester_id);
```

varqox commented on src/fs/metadata_log_operations/write.hh:
```diff
@@ -185,6 +185,13 @@ private:
+                        auto finished_cluster_id = offset_to_cluster_id(finished_cluster_offset, _metadata_log._cluster_size);
+                        if (buff_bytes_left == 0) {
+                            // Our disk_offset is pointing at the start of the next cluster
+                            finished_cluster_id--;
```
```suggestion
                            --finished_cluster_id;
```

wmitros committed:
fs: compaction: added calculation of metadata log and compacted metadata log sizes

Added new metadata log members:
    _log_cluster_count - which holds the number of clusters on disk that span current metadata log;
    _compacted_log_size - which holds sum of sizes of the structs that represent the same data using possibly little space;
Implemented calculations that update these values.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

wmitros committed:
fs: compaction: added initial cluster metadata

Created a new class - data_cluster_contents_info - which holds info about the data stored in a cluster.
Added new metadata log members which enable finding corresponding data_cluster_contents_info for every data cluster,
as well as discriminating the clusters that have been filled with data from ones that are in the middle of this process.
Implemented calculations that update the data stored in those metadata log fields.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -152,37 +192,53 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
+    auto& cc_info = cc_info_it->second;
+    cc_info.add_data(disk_offset, inode, {file_offset, file_offset + write_len});
+    if (created) {
+        _data_cluster_contents_info_map[cluster_id] = &cc_info;
```
We could put an assert here that this is creating an element indeed.
```suggestion
        auto& [_, inserted] = _data_cluster_contents_info_map.try_emplace(cluster_id, &cc_info);
        assert(inserted);
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -85,11 +85,11 @@ struct inode_info {
 
         // Deletes data vectors that are subset of @p data_range and cuts overlapping data vectors to make them
-        // not overlap. @p cut_data_vec_processor is called on each inode_data_vec (including parts of overlapped
-        // data vectors) that will be deleted
+        // not overlap. @p data_vec_cutting_callback is called with three parameters: an inode_data_vec that
+        // intersects with @p data_range and two data vector corresponding to its non-intersecting parts
```
I'm not sure if this is clear enough. Can someone else check this? @tropuq

tropuq replied:
I think it is missing information that inode_data_vec is called for every inode_data_vec that intersects data_range and explanation what are those `non-intersecting parts`. Probably it would be good to add information that those two non-intersecting parts can be non-empty only for inode_data_vecs on the edges of data_range.

wmitros replied:
The comments on lines 93,103-106 become redundant after including all this info, don't they?

varqox replied:
I think they are **not** redundant. Especially the second one, as it clearly illustrates the situation and which variable represents what.
I can see that you are trying to minimize the amount of code. Is that right? Please tell me why do you think these comments are redundant, because I don't see it.

wmitros replied:
> I think it is missing information that inode_data_vec is called for every inode_data_vec that intersects data_range

comment on line 93 doesn't speak about calling data_vec_cutting_callback but it does mention that we're working on every inode_data_vec that intersects data_range
> and explanation what are those `non-intersecting parts`. 

as you have noted, the second comment illustrates those 'non-intersecting parts'
> Probably it would be good to add information that those two non-intersecting parts can be non-empty only for inode_data_vecs on the edges of data_range.

this would be admittedly a new information 
>I can see that you are trying to minimize the amount of code. Is that right? Please tell me why do you think these comments are redundant, because I don't see it.

I'm actually trying to minimize the amount of remaining review, I've asked for clarification because I've already got reviews about comments being redundant and becoming outdated after implementing changes

tropuq replied:
Users of that function probably wouldn't like to read implementation of that function to know how it works so I think all necessary information should be outside the implementation.

wmitros replied:
I'm not saying the comment outside shouldn't be expanded. I'm asking if the comments inside should be removed while I'm expanding it

tropuq replied:
I think it depends on how you describe it outside the function. I would probably remove the comment from the inside if it was almost a copy-paste from the outside.

varqox replied:
> > I can see that you are trying to minimize the amount of code. Is that right? Please tell me why do you think these comments are redundant, because I don't see it.
> 
> I'm actually trying to minimize the amount of remaining review, I've asked for clarification because I've already got reviews about comments being redundant and becoming outdated after implementing changes

I thought that we want to optimize the output (high quality code that is ready to be merged) not the process to get to the output. I don't think that minimizing number of iterations of the review is the best choice. As the review is very fast now.
Regarding the comments, the comment above function declaration is to help the user of this function when he needs to use it. The comments inside implementation are to help the potential reader understand what is happening inside. One of aspect of why they are useful is their placement -- they are situated alongside code they refer to. And last but not least, comments in the implementation are used to minimize the time that reader needs to understand what is happening on such level that they know or are able to do what they wanted in the first place -- this can be e.g. editing this function, navigating through it to find relevant part (this is very important as it happens a lot, and in this process you don't read the whole implementation, you want as quickly as possible find what you want) or get a comprehensive understanding of what function does and how it does it.
Looking from that perspective, it make sense to leave these comments were they are because they help to achieve above goals. Maybe you don't need them, that is fine. But other people may find them very helpful, and I think it is very probable. As you don't write this code for yourself only, I suggest taking others perspective seriously into consideration, as people are very different and not everyone will understand what is going on there as quickly as you would.

wmitros replied:
@tropuq 
> I think it depends on how you describe it outside the function. I would probably remove the comment from the inside if it was almost a copy-paste from the outside.

That's what I was fearing would happen, but as we can see, there are uses for the inside comments even if they are quite similar

@varqox 
> I thought that we want to optimize the output (high quality code that is ready to be merged)

I've never disagreed with it and I don't know why does it need to be mentioned
> not the process to get to the output. 

As with the amount of code, minimizing the number of review iterations is hardly optimizing
> I don't think that minimizing number of iterations of the review is the best choice. As the review is very fast now.

While it's hard to disagree with that, it's also hard to agree that avoiding adding another review iteration has no benefits, given that we don't have unlimited time
> Regarding the comments, the comment above function declaration is to help the user of this function when he needs to use it. The comments inside implementation are to help the potential reader understand what is happening inside. One of aspect of why they are useful is their placement -- they are situated alongside code they refer to. And last but not least, comments in the implementation are used to minimize the time that reader needs to understand what is happening on such level that they know or are able to do what they wanted in the first place -- this can be e.g. editing this function, navigating through it to find relevant part (this is very important as it happens a lot, and in this process you don't read the whole implementation, you want as quickly as possible find what you want) or get a comprehensive understanding of what function does and how it does it.

Thank you for this fragment, because I believe it answers my question - why are the inside comments useful despite mostly repeating what's in the comment above the function.
> Looking from that perspective, it make sense to leave these comments were they are because they help to achieve above goals. Maybe you don't need them, that is fine. But other people may find them very helpful, and I think it is very probable. As you don't write this code for yourself only, I suggest taking others perspective seriously into consideration, as people are very different and not everyone will understand what is going on there as quickly as you would.

Isn't taking others perspective into consideration exactly what I'm doing by asking for that perspective instead of going ahead and deleting all the comments that I didn't find necessary?

All in all, I'm having an impression that I've been perceived as trying to reject any change or force my view while all I did was ask for a clarification that would help me write the high quality code that would not need further fixes.

varqox replied:
> > Looking from that perspective, it make sense to leave these comments were they are because they help to achieve above goals. Maybe you don't need them, that is fine. But other people may find them very helpful, and I think it is very probable. As you don't write this code for yourself only, I suggest taking others perspective seriously into consideration, as people are very different and not everyone will understand what is going on there as quickly as you would.
> 
> Isn't taking others perspective into consideration exactly what I'm doing by asking for that perspective instead of going ahead and deleting all the comments that I didn't find necessary?
> 
> All in all, I'm having an impression that I've been perceived as trying to reject any change or force my view while all I did was ask for a clarification that would help me write the high quality code that would not need further fixes.

Thank you for sharing that. I now see that I misunderstood your intentions and responded inadequately. I feel sad to see you felt sad because of what I wrote, I would have liked to be more aware of your needs.

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -152,37 +192,54 @@ void metadata_log::memory_only_delete_inode(inode_t inode) {
+    auto& cc_info = cc_info_it->second;
+    cc_info.add_data(disk_offset, inode, {file_offset, file_offset + write_len});
+    if (created) {
+        auto& [_, inserted] = _data_cluster_contents_info_map.try_emplace(cluster_id, &cc_info);
```
Compilation error.
